doc/elmalias.1
.if n \{\
.	ds ct "
.\}
.if t \{\
.	ds ct \\(co
.\}
.TH ELMALIAS 1L "Elm Version 2.5" "USENET Community Trust"
.SH NAME
elmalias - expand and display Elm address aliases
.SH SYNOPSIS
.B elmalias
[
.B \-adenrsuvV
] [
.B \-f
format ] [ name  ... ]
.sp
\fI$lib\fP/\fBprlong\fP
[ options ... ] < \f2file\fP
.SH DESCRIPTION
.I Elmalias
allows you to examine information about aliases.  The alias databases
are consulted for each
.I name
on the command line, and the alias value is displayed, one value per
line of output.  If
.I name
is not a known alias, then it is displayed unchanged.  If no
.I name
is given on the command line then all alias values in the databases
are displayed.  All of these actions are default behaviors that may
be modified by command line options.
.P
There are two possible alias databases, a user-specific alias file
(usually in
.IR $HOME/.elm/aliases )
and a system-wide alias file (usually in
.IR $lib/aliases ).
By default both files are searched, first the user-specific file and
then the system-wide file.  If an alias file does not exist then
.I elmalias
silently ignores it and continues on.  The
.B \-s
and
.B \-u
command line options (discussed shortly) can restrict what files are
files searched.
.P
There are two types of aliases, Person aliases and Group aliases.  A
Person alias expands to an individual address and a Group alias contains
a list of addresses, some of which may be aliases.  By default,
.I elmalias
displays the defined value of an alias, regardless of type.  The
.B \-e
command line option (discussed shortly) fully expands out Group aliases.
.P
The default output produced is just the ``Address'' information for
each
.I name
on the command line.  A number of command line options (discussed
shortly) may be used to select different output information or specify
a format of your choosing.  The following information is maintained
in the alias databases, and may be accessed by the
.I elmalias
utility:
.nf

	\(bu Alias (the alias name)
	\(bu Last Name
	\(bu Name (the user's full name)
	\(bu Comment (not displayed in mail headers)
	\(bu Address (the alias value)
	\(bu Type (Person, Group, or Unknown)

.fi
.P
When the
.I name
specified on the command line is not a known alias, both the ``Alias''
and ``Address'' information will display as the
.IR name ,
the ``Type'' information will display as ``Unknown'', and all other items
will be blank.
.P
The available command line options are:
.P
.\"
.\"
.\"
.IP \fB\-a\fP
Selects an alternative output format.  The ``Alias'' information is
displayed in addition to and preceding the usual ``Address'' information.
.\"
.\"
.\"
.IP \fB\-d\fP
Turns debugging on. Has no effect unless
.I elmalias
was compiled with debugging enabled.
.\"
.\"
.\"
.IP \fB\-e\fP
Tells
.I elmalias
to fully expand group aliases.  This option can be used only when
.IR name s
are given on the command line.
.\"
.\"
.\"
.IP \fB\-f\fP
The output produced will be in the specified
.IR format .
Format specifications will be discussed shortly.
.\"
.\"
.\"
.IP \fB\-n\fP
Selects an alternative output format.  The ``Name'' information, if
it exists, is displayed in addition to and following the usual
``Address'' information.  The ``Name'' will be enclosed in parenthesis.
.\"
.\"
.\"
.IP \fB\-r\fP
If a specified
.I name
does not correspond to a known alias then
.I elmalias
will display a message and exit with a nonzero status.
.\"
.\"
.\"
.IP \fB\-s\fP
Normally
.I elmalias
examines both the user-specific and system-wide alias files.  This option
requests that
.I elmalias
use the system-wide alias file, and unless the
.B \-u
option is also specified the user-specific alias file will be ignored.
.\"
.\"
.\"
.IP \fB\-u\fP
Normally
.I elmalias
examines both the user-specific and system-wide alias files.  This option
requests that
.I elmalias
use the user-specific alias file, and unless the
.B \-s
option is also specified the system-wide alias file will be ignored.
.\"
.\"
.\"
.\"
.\"
.\"
.IP \fB\-v\fP
Selects a verbose output format.  Both the ``Alias'' and the ``Name''
information will be displayed in addition to the ``Address'' information.
The ``Alias'' information will precede the ``Address'' information,
and the ``Name'' information will be enclosed in parenthesis and follow it.
If there is no ``Name'' information for this alias then that portion of the
output will be ignored.
.\"
.\"
.\"
.IP \fB\-V\fP
Selects a very verbose, multi-line output format.  Displays all available
information on aliases.
.\"
.\"
.\"
.P
The output produced by
.I elmalias
is fully customizable with the
.B \-f
option.  The
.I format
string uses a syntax similar to
.I date(1)
(System V version) and
.IR printf(3) .
The following field descriptors may be used in
.I format
specifications:
.nf

	%a	Alias (the alias name)
	%l	Last Name
	%n	Name (the user's full name)
	%c	Comment (not displayed in mail headers)
	%v	Address (the alias value)
	%t	Type (Person, Group, or Unknown)

.fi
Field widths in a
.I [-][m][.n]
format (again, similar to
.IR printf(3) )
may also be used.  For example, "%-20.20a" means print the ``Alias''
information left justified in a field twenty characters long, with
the value truncated to twenty characters.
.P
The following special character sequences are also recognized in
format specifications:
.nf

	\\b	A backspace.
	\\f	A formfeed.
	\\n	A newline.
	\\r	A return.
	\\t	A tab.
	\\\fIc\fP	Literal character ``\fIc\fP''.

.fi
There is a very simplistic conditional evaluation mechanism that may
be used in format specifications.  The conditional text should be
surrounded by question marks, and a single character that specifies
the condition immediately follows the first question mark.  The
condition characters correspond to the ``%'' field specifier characters,
and the condition is true if the corresponding alias information is
defined and nonempty.  For example, if you want to display the ``Name''
information surrounded by parenthesis, but omit it if the information
is not available, you may use ``?n(%n)?'' in the format specification.
.P
The command line switches that select an alternative format correspond
to the following format specifiers.
.nf

	\fIdefault\fP	"%v"
	\-a	"%-20.20a %v"
	\-n	"%v?n (%n)?"
	\-v	"%-20.20a %v?n (%n)?"
	\-V	"Alias:\\t\\t%a\\n\\
		  Address:\\t%v\\n\\
		  Type:\\t\\t%t\\n\\
		?n  Name:\\t\\t%n\\n?\\
		?l  Last Name:\\t%l\\n?\\
		?c  Comment:\\t%c\\n?"

.fi
.\"
.\"
.\"
.P
The
.I prlong
utility formats long amounts of data, folding across multiple lines.
It is useful to reformat output of
.IR elmalias .
.I Prlong
reads data from its standard input, one line at a time, and tries to
place as much information as possible on each output line.  A field
seperator, by default a single space, seperates each input record in
the output lines.  Every output line is preceded by a leader field.
By default the leader of the first output line is an empty string,
and the leader for all subsequent output lines is a single tab.
.I prlong
will never split an input record.  If an input record exceeds the
maximum output line length, it will appear unmodified on an output
line all by itself.
.P
The
following options may be used to modify the
.I prlong
behavior.
.IP "\fB\-w\fP \fIwidth\fP" 12
Constrains output lines to
.I width
columns in length (not counting an appended newline character).  The
default is 78 columns.  The width calculation assumes that tabstops
occur every eight spaces.
.IP "\fB\-f\fP \fIstring\fP" 12
Seperates each input record with the indicated
.I string
when displayed to the output.  The default is a single space.
.IP "\fB\-1\fP \fIstring\fP" 12
Specifies the leader
.I string
used for the first line of output.  (This option is the digit ``one''.)
The default is an empty string.
.IP "\fB\-l\fP \fIstring\fP" 12
Specifies the leader
.I string
for all subsequent lines of output.  (This option is a lower-case
``ell''.) The default is a single tab.
.SH EXAMPLES
.P
Consider an
.I aliases.text
file that contains:
.nf
.ft CW

	friends = List of Friends = tom, dick, harry
	tom = Tom Smith = sleepy!tom
	dick = Dick Jones = dopey!dick
	harry = = grumpy!harry

.ft R
.fi
Below are shown some example commands and the output produced.
.nf
.ft CW

	$ \fIelmalias friends\fP
	tom,dick,harry
.sp 0.2v
	$ \fIelmalias mike\fP
	mike
.sp 0.2v
	$ \fIelmalias -r mike\fP
	elmalias: "mike" is not a known alias
.sp 0.2v
	$ \fIelmalias -n friends\fP
	tom,dick,harry (List of Friends)
.sp 0.2v
	$ \fIelmalias -a friends\fP
	friends              tom,dick,harry
.sp 0.2v
	$ \fIelmalias -V friends\fP
	Alias:          friends
	  Address:      tom,dick,harry
	  Type:         Group
	  Name:         List of Friends
	  Last Name:    List of Friends
.sp 0.2v
	$ \fIelmalias -e friends\fP
	tom@sleepy.acme.com
	dick@dopey.acme.com
	harry@grumpy.acme.com
.sp 0.2v
	$ \fIelmalias -ve friends\fP
	tom                  tom@sleepy.acme.com (Tom Smith)
	dick                 dick@dopey.acme.com (Dick Jones)
	harry                harry@grumpy.acme.com
.sp 0.2v
	$ \fIelmalias -f "alias %a is \\"%v\\" ?n(%n)?" -e friends\fP
	alias tom is "tom@sleepy.acme.com" (Tom Smith)
	alias dick is "dick@dopey.acme.com" (Dick Jones)
	alias harry is "harry@grumpy.acme.com" 
.sp 0.2v
	$ \fIelmalias -en friends | /usr/lib/elm/prlong -w40\fP
	tom@sleepy.acme.com (Tom Smith) 
	        dick@dopey.acme.com (Dick Jones) 
	        harry@grumpy.acme.com
.sp 0.2v
	$ \fIelmalias -en friends | /usr/lib/elm/prlong -1 "To: " -f ", " -w40\fP
	To: tom@sleepy.acme.com (Tom Smith), 
	        dick@dopey.acme.com (Dick Jones), 
	        harry@grumpy.acme.com

.ft R
.fi
.RE
.P
The
.IR checkalias (1L)
and
.IR listalias (1L)
scripts distributed with the Elm package provide further examples
of the
.I elmalias
and
.I prlong
utilities.
.SH AUTHOR
Elm Development Group
.SH SEE ALSO
checkalias(1L),
elm(1L),
listalias(1L),
newalias(1L)
.SH BUGS
There is no centralized, comprehensive interface to the Elm 2.4 alias
system, so every program in the Elm suite that uses aliases implements
its own interfaces.  It's possible for inconsistencies to creep in
between what
.I elmalias
says and what some Elm utility does.  If you find an inconsistency
please report it to us!
.P
Boy, there sure are a lot of command line switches in this thing.
.P
Note that the precedence of aliases is user file then system
file.  This means that a user can 'overload' an alias by having
one defined in the system file also defined in theirs.  This
shouldn't turn out to be a problem, but is something for
the system administrator to keep in mind when creating the
system alias file.
.SH BUG REPORTS TO
Bill Pemberton  flash@virginia.edu
.SH COPYRIGHTS
\fB\*(ct\fRCopyright 1993-1995 by The USENET Community Trust


Changes
Changes in 2.5 PL8

- fixed a hang when hitting certain 8-bit chars in the From line

- fixed a file descriptor leak (from Andreas Luik)

- removed all Expires: processing


Changes in 2.5 PL7

- a few display issues fixed (from Kenneth Reek)

- buffer overflow in frm fixed

Changes in 2.5 PL6

- Fixed message-id in the In-Reply-To header.

- Actually show the -r option in the usage

- Whacked a few more places where we might follow a symlink in tmpdir

- Fastmail didn't grok commas correctly.

- Change lock.c to not use errno unless there's an errno

Changes in 2.5 PL5

- A few more buffer overruns fixed.

Changes in 2.5 PL4

- little bug in aliasdb.c (from Scot W. Hetzel <scot@genroco.com>)

- Message-Id: could overflow (from Darin Wayrynen <darin@deru.net>)

- Fixed the permissions setting problem with the edit mailbox function

- Added a ResizeScreen() call after show_msg() calls to hopefully deal wit
  SIGWINCH problems (inspired from Yuke Zhuge <zhuge@stone.pdx.informix.com>)
  
- Fixed problems under SunOS 4.1.3 (from Paul J. Menchini <mench@mench.com>)

- Fixed content encoding (from David Woolley  <david@djwhome.demon.co.uk>)

- Attempt to stop crashes with long recipient list (from Gilles Detillieux 
  <grdetil@scrc.umanitoba.ca>)
  

Changes in 2.5 PL3

- Fixed the declaration of outbound_from_name (from Richard Jackson 
  <rjackson@gmu.edu>)

- Fixed the leapyear macro (from Gerhard Niklasch)
  
- Fixed editing of a long CC line (from Derek Godfrey <djg@sjc.ascend.com>)

- Mail that lands in the sentmail with attachments now has the 
  Content-disposition header text added to the message (from Dave 
  Spensley)
  
- Made allto a bit longer (suggested by Dave Spensley 
  <davesp@typhoon.xnet.com>)

- Fixed Configure to find (again) fcntl on Linux

- can_access will now allow you to write to a char device (/dev/null)

Changes in 2.5 PL2

- Fixed (again, maybe?) savebyalias from Adam Benjamin <aebenjam@opentext.com>

- Fixed bug were an alias delete would crash elm if the mailbox was empty

Changes in 2.5 PL1

- Fixed savebyalias from Michael D Hughes <mdh@logcabin.woods.bridge.com> 
  and Robert Schwartz <rjs@elt.com>
  
- Changes for 8 bit clean from H.J. Lu <hjl@varesearch.com>

Changes in 2.5 PL0

- One minor change to make an easier compile on SparcLinux.

Changes in 2.5 PL0pre8

- Y2K stuff from Ed Symanzik <zik@msu.edu>

Changes in 2.5 PL0pre7

- minor change to deal with egcs

Changes in 2.5 PL0pre6

- newalias nows does a geteuid() instead of getuid().  This is to lets 
  SysAdmins do "su <user> -c 'newalias'"
  

Changes in 2.5 PL0pre5

- Fix for GECOS entries with comma-delimited sub-fields from Chip 
  Rosenthal <chip@unicom.com>.

- Fix? for keys getting screw up after a ^Z from Chip Rosenthal
  <chip@unicom.com>.


Changes in 2.5 PL0pre4

- A few changes for BSD from Chip Rosenthal <chip@unicom.com>.

- A few changes to Configure suggested by Chip Rosenthal <chip@unicom.com>.


Changes in 2.5 PL0pre3

- Configure didn't see fcntl locking on Linux.  Fix from Sean McDermott 
  <spm@intel.com>


Changes in 2.5 beta 3

- Fixed a problem where elm would try to close a file twice when
  sending an attachment (from reichler@ews.uiuc.edu)

- Added docs for environment variables to the man page (from Eric
  Haszlakiewicz <nimenees@uaccess.net>)

Changes in 2.5 beta 2

- IncomingFolders no longer set Magic Mode on for everything

- Configure should now deal with Linux.

- Make //pattern honor current limit (from Keith Neufeld<neufeld@pvi.org>)

- year 2000 fix.....?

- Building on Solaris requires /usr/ccs/bin to be in the PATH

Changes in 2.5 beta 1

- b)ounce honor DONT_ADD_FROM (from Keith Neufeld<neufeld@pvi.org>)

- removed [encode] display and reply (from Keith Neufeld<neufeld@pvi.org>)

- make right/left arrows work in the file browser the same as in the
  message index. (from Keith Neufeld<neufeld@pvi.org>)

- fixed a bug in the incomingfolders code (it wouldn't save) (from Bob Schwartz
  <rjs@elt.com>)
  
- fixed the meaning of noheader for forwards (from Keith Neufeld<neufeld@pvi.org>)



Changes in 2.5 alpha 11

- fixed a bug in the alias code (thanks to Keith Neufeld<neufeld@pvi.org>)

- fixed the generation of temp file names in newmbox.c

- allowed longer Cc: lines.  This was due to a crash with a group reply.

- added incomingfolders to the elmrc.  This is a list of folders that will
  automatically be opened with "magic mode" on. (inspired by a patch from
  Cristophe Kalt <kalt@ensta.fr>)


Changes in 2.5 alpha 10

- Press u on a message that is read but not deleted will mark the message as
  unread (from Larry Schwimmer <rosebud@cyclone.Stanford.EDU>).

- mmencode.1 man page is now installed with others (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- Magic mode stuff is now in help and NLS (from Gregory Margo
  <gmargo@newton.vip.best.com>).

- allow elimination of aliases from larger aliases (To: bigalias -subalias).
  Based on incomplete patch by Waldi Ravens <waldi@moacs.indiv.nl.net>.  Also
  removes redundant declarations of get_alias_address(), which is declared in
  elm_globals.h (from Keith Neufeld<neufeld@pvi.org>)

- On an entry field that normally clears to be retyped from scratch as soon
  as the user presses a key (Subject: on header edit screen), allow right
  arrow and delete to begin editing the existing default answer. (from Keith
  Neufeld<neufeld@pvi.org>)

- Add a warning to Configure about disabling shell escapes (from Keith
  Neufeld<neufeld@pvi.org>)

- Eliminate complaints in doc/Makefile about duplicate makefile entries if
  mansrc and catmansrc are both "none" (from Keith Neufeld<neufeld@pvi.org>)

- Allow entry of uppercase commands on header display/edit screens, since
  that's how they're parenthesized on the display.  (from Bob Schwartz
  <rjs@elt.com> and Keith Neufeld <neufeld@pvi.org>

- Make 2+ -line error messages write over more of the screen instead of

utils/elmalias.c


/*******************************************************************************
 *  The Elm Mail System  -  $Revision: 1.4 $
 *
 *                      Copyright (c) 1988-1995 USENET Community Trust
 * 			Copyright (c) 1986,1987 Dave Taylor
 *******************************************************************************
 * Bug reports, patches, comments, suggestions should be sent to:
 *
 *      Bill Pemberton, Elm Coordinator
 *      flash@virginia.edu
 *
 *******************************************************************************
 * $Log: elmalias.c,v $
 * Revision 1.4  1996/03/14  17:30:07  wfp5p
 * Alpha 9
 *
 * Revision 1.3  1996/03/13  14:38:05  wfp5p
 * Alpha 9 before Chip's big changes
 *
 * Revision 1.2  1995/04/20  21:02:05  wfp5p
 * Added the showreply feature and emacs key bindings.
 *
 * Revision 1.1.1.1  1995/04/19  20:38:40  wfp5p
 * Initial import of elm 2.4 PL0 as base for elm 2.5.
 *
 ******************************************************************************/

/*
 * elmalias - Elm alias database access.
 *
 * This utility will display information from the Elm user and system
 * alias databases.  It will expand alias names specified on the command
 * line and display the alias value.  If an item on the command line is
 * not a valid alias, then it's value is returned.  If no names are
 * specified on the command line then all alises are displayed.  Output
 * is formatted by either a user-defined format specifier or one of a
 * number of predefined specifications.  The default output format
 * displays just the alias value.
 *
 * The format specifiers are:
 *
 *	%a	alias (the alias name)
 *	%l	last_name
 *	%n	name
 *	%c	comment
 *	%v	address (the alias value)
 *	%t	type (Person, Group, or Unknown)
 *
 * In the case that a command line argument does not correspond to
 * an alias, then %a and %v will evaluate to the argument value and
 * all other items will be empty.
 *
 * Printf-like field widths may be used, e.g. "%-20.20a".  Conditionals
 * are bracketed by question marks.  For example "?n(%n)?" means format
 * "(%n)" if the "name" is defined, otherwise produce no output.  Some
 * backslash sequences (e.g. "\t" == tab) are recognized.  Any other
 * characters are displayed as is.
 *
 * Example:
 *	$ elmalias -f "alias %a is address \"%v\" ?n(%n)?" chip
 *	alias chip is address "chip@chinacat.unicom.com" (Chip Rosenthal)
 *
 * Synopsis:
 *	elmalias [-adensuvV] [-f format] [alias ...]
 *
 * Options:
 *	-a	Display alias name.  Equivalent to -f "%-20.20a %v".
 *	-d	Turn debugging on. Only useful if DEBUG is defined.
 *	-e	Fully expand alias values.
 *	-f fmt	User-specified output format.
 *	-n	Display name.  Equivalent to -f "%v?n (%n)?".
 *	-r	Complain about arguments that are not valid aliases.
 *	-s	Consult system-wide alias file.
 *	-u	Consult user-specific alias file.
 *	-v	Verbose output.  Equivalent to -f "%-20.20a %v?n (%n)?".
 *	-V	Babble about *everything* we known about.
 *
 * If neither -s or -u are specified, then the default is to search
 * both alias files.
 */

#define INTERN
#include "elm_defs.h"
#include "s_elmalias.h"
#include "ndbz.h"

/*
 * Maximum number of alias files we can consult.
 */
#define MAXDB		2	/* user and system alias files		*/

char *Progname;

/*
 * "aliasdb" library routines.
 */
extern struct alias_rec *fetch_alias();
extern char *next_addr_in_list();

/*
 * Local procedures.
 */
DBZ *open_user_aliases(), *open_system_aliases();
struct alias_rec *make_dummy_rec();
void exp_print_alias(), print_alias();
char *sel_alias_mem();


void usage_error()
{
    fprintf(stderr, catgets(elm_msg_cat, ElmaliasSet, ElmaliasUsage,
	"usage: %s [-adenrsuvV] [-f format] [alias ...]\n"), Progname);
    exit(1);
    /*NOTREACHED*/
}


main(argc, argv)
int argc;
char *argv[];
{
    char *out_fmt;		/* output printing format		*/
    int do_user_alias;		/* TRUE to examine user alias file	*/
    int do_system_alias;	/* TRUE to examine system alias file	*/
    int do_expand;		/* TRUE to recursively expand aliases	*/
    int do_complain;		/* TRUE to insist args are valid aliases*/
    int numdb;			/* number of alias files to consult	*/
    DBZ *dblist[MAXDB+1];	/* NULL terminated list of files	*/
    struct alias_rec *ar;	/* scratch ptr to expansion of curr name*/
    int d, i;
    extern int optind;
    extern char *optarg;

    /*
     * Initialize.
     */
    initialize_common();
    Progname = argv[0];		/* program name for diag messages	*/
    do_user_alias = FALSE;	/* indicate the user hasn't selected	*/
    do_system_alias = FALSE;	/*   any alias files (yet)		*/
    do_expand = FALSE;		/* do not recursively expand groups	*/
    do_complain = FALSE;	/* allow non-aliases on cmd line	*/
    out_fmt = "%v";		/* default is to just show alias value	*/
    numdb = 0;			/* no alias files opened yet		*/

    /*
     * Crack command line options.
     */
    while ((i = getopt(argc, argv, "adef:nrsuvV")) != EOF) {
	switch (i) {
	case 'a':			/* show alias name and value	*/
	    out_fmt = "%-20.20a %v";
	    break;
	case 'd':
#ifdef DEBUG
	    debug = 10;
#endif
	    break;
	case 'e':			/* recursively expand aliases	*/
	    do_expand = TRUE;
	    break;
	case 'f':			/* user-specified format	*/
	    out_fmt = optarg;
	    break;
	case 'n':			/* show alias value and fullname*/
	    out_fmt = "%v?n (%n)?";
	    break;
	case 'r':			/* insist args are valid aliases*/
	    do_complain = TRUE;
	    break;
	case 's':			/* use system-wide alias file	*/
	    do_system_alias = TRUE;
	    break;
	case 'u':			/* use per-user alias file	*/
	    do_user_alias = TRUE;
	    break;
	case 'v':			/* verbose output format	*/
	    out_fmt = "%-20.20a %v?n (%n)?";
	    break;
	case 'V':			/* show the user's life story	*/
	    out_fmt = "\
Alias:\t\t%a\n\
  Address:\t%v\n\
  Type:\t\t%t\n\
?n  Name:\t\t%n\n?\
?l  Last Name:\t%l\n?\
?c  Comment:\t%c\n?\
";
	    break;
	default:
	    usage_error();
	    /*NOTREACHED*/
	}
    }

    /*
     * If user didn't request specific alias files then use them all.
     */
    if (!do_system_alias && !do_user_alias)
	do_system_alias = do_user_alias = TRUE;

    /*
     * Open up the alias files we need to access, in the order of priority.
     */
    if (do_user_alias && (dblist[numdb] = open_user_aliases()) != NULL)
	++numdb;
    if (do_system_alias && (dblist[numdb] = open_system_aliases()) != NULL)
	++numdb;
    dblist[numdb] = NULL;

    /*
     * If no names specified on command line then dump all alias files..
     */
    if (optind == argc) {
	if (do_expand) {
	    fprintf(stderr, catgets(elm_msg_cat, ElmaliasSet,
		ElmaliasCannotSpecifyExpand,
		"%s: cannot specify \"-e\" when dumping all aliases\n"),
		Progname);
	    exit(1);
	}
	for (d = 0 ; dblist[d] != NULL ; ++d) {
	    /* assumes file pointer left at first key immediately after open */
	    while ((ar = fetch_alias(dblist[d], (char *)NULL)) != NULL) {
		print_alias(out_fmt, ar);
		(void) free((malloc_t)ar);
	    }
	}
	exit(0);
    }

    /*
     * Expand each name on the command line.
     */
    for (i = optind ; i < argc ; ++i) {

	/*
	 * Try each of the alias files for a match.
	 */
	ar = NULL;
	for (d = 0 ; dblist[d] != NULL ; ++d) {
	    if ((ar = fetch_alias(dblist[d], argv[i])) != NULL)
		break;
	}

	/*
	 * Print the result.
	 */
	if (ar == NULL) {
	    if (do_complain) {
		fprintf(stderr, catgets(elm_msg_cat, ElmaliasSet,
		    ElmaliasUnknownAlias, "%s: \"%s\" is not a known alias\n"),
		    Progname, argv[i]);
		exit(1);
	    }
	    ar = make_dummy_rec(argv[i]);
	    print_alias(out_fmt, ar);
	} else if (do_expand && (ar->type & GROUP)) {
	    exp_print_alias(dblist, out_fmt, ar);
	} else {
	    print_alias(out_fmt, ar);
	}

	(void) free((malloc_t)ar);

    }

    exit(0);
    /*NOTREACHED*/
}


DBZ *open_system_aliases()
{
    return dbz_open(system_data_file, O_RDONLY, 0);
}


DBZ *open_user_aliases()
{
    char fname[SLEN];
    sprintf(fname, "%s/%s", user_home, ALIAS_DATA);
    return dbz_open(fname, O_RDONLY, 0);
}


/*
 * Cobble up an alias record structure to hold some address info.
 */
struct alias_rec *make_dummy_rec(val)
char *val;
{
    struct alias_rec *ar;
    ar = (struct alias_rec *) safe_malloc(sizeof(struct alias_rec));
    ar->status = 0;
    ar->alias = val;
    ar->last_name = "";
    ar->name = "";
    ar->comment = "";
    ar->address = val;
    ar->type = 0;
    ar->length = 0;
    return ar;
}


/*
 * Recursively expand out a list of addresses and print the expansions.
 */
void exp_print_alias(dblist, fmt, ar)
DBZ *dblist[];
char *fmt;
struct alias_rec *ar;
{
    char *abuf;		/* list of addresses we can scribble upon	*/
    char *acurr;	/* pointer to current address within "abuf"	*/
    char *anext;	/* pointer to next address within "abuf"	*/
    struct alias_rec *ar0;
    int d;		/* dblist index					*/

    /*
     * Create a copy of this address we can scribble upon.
     */
    anext = abuf = safe_strdup(ar->address);
#ifdef lint
    *abuf = '\0'; /* shutup set but not used complaint */
#endif

    /*
     * Go through all of the addresses and expand them out.
     */
    while ((acurr = next_addr_in_list(&anext)) != NULL) {
	for (d = 0 ; dblist[d] != NULL ; ++d) {
	    if ((ar0 = fetch_alias(dblist[d], acurr)) != NULL)
		break;
	}
	if (ar0 == NULL)
	    ar0 = make_dummy_rec(acurr);
	if (ar0->type & GROUP)
	    exp_print_alias(dblist, fmt, ar0);
	else
	    print_alias(fmt, ar0);
	(void) free((malloc_t)ar0);
    }

    (void) free((malloc_t)abuf);
}


/*
 * Print out alias information according to a format specification.
 */
void print_alias(fmt, ar)
char *fmt;
struct alias_rec *ar;
{
    char pfmt[64];		/* buffer to hold "%m.ns" formats	*/
    int in_conditional;		/* TRUE if in middle of cond expression	*/
    int print_enab;		/* TRUE if OK to print output		*/
    char *s;
    int n, c;

    print_enab = TRUE;
    in_conditional = FALSE;

    while (*fmt != '\0') {

	switch (*fmt) {

	/*
	 * Formatted output.
	 */
	case '%':

	    /*
	     * Extract the "%m.n" portion of the format.
	     */
	    pfmt[0] = *fmt++;
	    n = 1;
	    while (index("-.0123456789", *fmt) != NULL) {
		if (n < sizeof(pfmt)-2)
		    pfmt[n++] = *fmt;
		++fmt;
	    }
	    pfmt[n++] = 's';
	    pfmt[n] = '\0';

	    /*
	     * Determine what we are printing.
	     */
	    if ((s = sel_alias_mem(ar, *fmt)) == NULL) {
		s = catgets(elm_msg_cat, ElmaliasSet, ElmaliasIllegalFmtChar,
		    "<illegal format char>");
	    }

	    /*
	     * Print out the formatted field.
	     */
	    if (print_enab)
		printf(pfmt, s);
	    break;

	/*
	 * Conditional printing.
	 */
	case '?':
	    if (in_conditional) {
		in_conditional = FALSE;
		print_enab = TRUE;
	    } else {
		in_conditional = TRUE;
		s = sel_alias_mem(ar, *++fmt);
		print_enab = (s != NULL && *s != '\0');
	    }
	    break;
	
	/*
	 * Backslash escapes.
	 */
	case '\\':
	    switch (*++fmt) {
		case 'b':  c = '\b'; break;
		case 'f':  c = '\f'; break;
		case 'n':  c = '\n'; break;
		case 'r':  c = '\r'; break;
		case 't':  c = '\t'; break;
		default:   c = *fmt; break;
	    }
	    if (print_enab && c != '\0')
		putchar(c);
	    break;

	/*
	 * Non-special character to print.
	 */
	default:
	    if (print_enab)
		putchar(*fmt);
	    break;

	}

	if (*fmt != '\0')
	    ++fmt;

    }

    putchar('\n');
}


/*
 * Select a member of the alias record structure.
 */
char *sel_alias_mem(ar, sel)
struct alias_rec *ar;
int sel;
{
    switch (sel) {
    case 'a':
	return ar->alias;
    case 'l':
	return ar->last_name;
    case 'n':
	return ar->name;
    case 'c':
	return ar->comment;
    case 'v':
	return ar->address;
    case 't':
	switch (ar->type & (PERSON|GROUP)) {
	case PERSON:
	    return catgets(elm_msg_cat, ElmaliasSet,
		ElmaliasTypePerson, "Person");
	case GROUP:
	    return catgets(elm_msg_cat, ElmaliasSet,
		ElmaliasTypeGroup, "Group");
	default:
	    return catgets(elm_msg_cat, ElmaliasSet,
		ElmaliasTypeUnknown, "Unknown");
	}
    default:
	return (char *) NULL;
    }
    /*NOTREACHED*/
}

